src/stats/probs.rs

```diff
@@ -0,0 +1,352 @@
+        Clone,
+        Debug
+    )]
+    /// A newtype for PHRED-scale probabilities.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// #[macro_use]
+    /// extern crate approx;
+    /// # extern crate bio;
+    /// # fn main() {
+    /// use bio::stats::{PHREDProb, Prob};
+    ///
+    /// let p = PHREDProb::from(Prob(0.5));
+    ///
+    /// assert_relative_eq!(*Prob::from(p), *Prob(0.5));
+    /// # }
+    /// ```
+    pub struct PHREDProb(pub f64);
+}
+
+
+/// Iterator returned by scans over logprobs.
+pub type ScanIter<I> = iter::Scan<<I as IntoIterator>::IntoIter, LogProb, fn(&mut LogProb, LogProb) -> Option<LogProb>>;
+
+
+impl LogProb {
+    /// Log-space representation of Pr=0
+    pub fn ln_zero() -> LogProb {
+        LogProb(f64::NEG_INFINITY)
+    }
+
+    /// Log-space representation of Pr=1
+    pub fn ln_one() -> LogProb {
+        LogProb(0.0)
+    }
+
+    /// Numerically stable calculation of 1 - p in log-space.
+    pub fn ln_one_minus_exp(&self) -> LogProb {
+        LogProb(ln_1m_exp(**self))
+    }
+
+    /// Numerically stable sum of probabilities in log-space.
+    pub fn ln_sum_exp(probs: &[LogProb]) -> LogProb {
+        if probs.is_empty() {
+            Self::ln_zero()
+        } else {
+            let mut pmax = probs[0];
+            let mut imax = 0;
+            for (i, &p) in probs.iter().enumerate().skip(1) {
+                if p > pmax {
+                    pmax = p;
+                    imax = i;
+                }
+            }
+            if pmax == Self::ln_zero() {
+                Self::ln_zero()
+            } else if *pmax == f64::INFINITY {
+                LogProb(f64::INFINITY)
+            } else {
+                // TODO use sum() once it has been stabilized: .sum::<usize>()
+                pmax + LogProb(
+                    (probs.iter()
+                          .enumerate()
+                          .filter_map(|(i, p)| {
+                              if i == imax {
+                                  None
+                              } else {
+                                  Some((p - pmax).exp())
+                              }
+                          })
+                          .fold(0.0, |s, e| s + e)
+                     ).ln_1p()
+                 )
+            }
+        }
+    }
+
+    /// Numerically stable addition probabilities in log-space.
+    pub fn ln_add_exp(self, other: LogProb) -> LogProb {
+        let (mut p0, mut p1) = (self, other);
+        if p1 > p0 {
+            mem::swap(&mut p0, &mut p1);
+        }
+        if p0 == Self::ln_zero() {
+            Self::ln_zero()
+        } else if *p0 == f64::INFINITY {
+            LogProb(f64::INFINITY)
+        } else {
+            p0 + LogProb((p1 - p0).exp().ln_1p())
+        }
+    }
+
+    /// Numerically stable subtraction of probabilities in log-space.
+    pub fn ln_sub_exp(self, other: LogProb) -> LogProb {
+        let (p0, p1) = (self, other);
+        assert!(p0 >= p1,
```

Comment:
Would a Result<LogProb, BioError> return type be more appropriate here? How likely is it for a consumer of this module to accidentally subtract a larger probability from a smaller one?